### ISSUE TITLE
Generalizing Bosch alarm support

### ISSUE THREAD

**Issue by mag1024** [sanjay900/solution3000-home-assistant#6]
Hi, I am the developer of a python [library](https://github.com/mag1024/bosch-alarm-mode2) and a corresponding [HA integration](https://github.com/mag1024/bosch-alarm-homeassistant) for interfacing with Bosch "Mode 2" alarm panels.
I was recently made aware of your project via [this](https://github.com/mag1024/bosch-alarm-mode2/issues/1) discussion. It appears that there are quite a few differences in how the different panels implement the protocol, and our corresponding projects only work with a single panel model. What do you think about somehow merging our projects into a single, more general implementation?

**Comment by sanjay900** [comment]
Funnily enough, I'm going to be rewriting this because I'm being given some
hardware with subscription capabilities on the 3000, so we could definitely
look into merging our codebases

**Comment by sanjay900** [comment]
I'm actually working with Bosch on this, so i have obtained documentation and was planning on doing this all properly once i get this hardware from them as well. I've finally obtained a copy of the mode 2 documentation myself, but im waiting on this hardware to arrive before i do much else. Which panel have you got?

It would make sense for us to merge our efforts, especially since i was going to need to implement subscriptions anyway, just not really sure what the best method to approach that is.

**Comment by mag1024** [comment]
Cool!
I have the B8512G, and getting push events to work was one of my core goals with this project, since I wanted to be able to use the motion sensors wired to the alarm to trigger lights, and polling wouldn't be fast enough for that. To this end, the library is written to be fully async, which seems to work pretty well. Here's the [callback](https://github.com/mag1024/bosch-alarm-mode2/blob/a98be6437ac2095aee39f9c9e0796e4c6ee8851b/bosch_alarm_mode2/panel.py#L311) that processes subscription updates.

I have Bosch's "Intrusion Integration Protocol Development Documentation", and it seems to call out _some_ of the differences between the panels, but clearly not all. For example, it mentions the different "arm" modes used by Solution panels, but not the fact that it requires different authentication (as discussed in [mag1024/bosch-alarm-mode2#1](https://github.com/mag1024/bosch-alarm-mode2/issues/1) -- this would probably be the first thing to fix). So ultimately, the only reliable way to develop support for particular models is to test on physical hardware.

**Comment by sanjay900** [comment]
I've got a draft version of that document with more info but haven't
actually read through to see if it gives any more info on the login process
or not, and I'm waiting for my development hardware to arrive to see if
anything changed with the new firmware on that front too

**Comment by sanjay900** [comment]
My new solution panel with push event support should be arriving this week, how do we want to work on merging our projects together?

Note that bosch wants to eventually get this upstreamed into home assistant proper as an offical integration as well.

**Comment by mag1024** [comment]
Being as objective as possible, I do think bosch-alarm-mode2 is a solid foundation to build upon. It's built natively for async, and is hopefully laid out well enough that it should be easy to follow. I would suggest we start by getting the authentication to work (make basic info happen first, then use different authentication mechanisms conditionally on the panel model), and see what else doesn't work. From there maybe you could look at various panel-specific issues, and I could pick up the implementation of some of the missing functionality (the library is relatively bare-bones right now).

It's great to hear that we have Bosch's official blessing! Given the aggressive legalese at the beginning of the development documentation, it wasn't clear to me how open-source friendly they were.

**Comment by mag1024** [comment]
BTW, the fastest way to iterate on authentication/receiving events is by running [this](https://github.com/mag1024/bosch-alarm-mode2/blob/master/bin/alarm.py) standalone tool (need to run it inside `poetry shell`).

**Comment by sanjay900** [comment]
Do note that the solution panels don't support async with their current firmware as well, so you will need to support a polled version as well. That is why my implementation is not async based.

There is a command you can send to check if async is supported or not, so we can make use of that instead of needing to hardcoded in panel support, since  async is coming to the solutions series
I will have both a async enabled panel with a development firmware, and a normal panel available to me soon, just waiting for the test panel to arrive in the post at the moment, it should be here this week

I do like the idea of building upon `bosch-alarm-mode2`, but how hard would it be to get a polled implementation into that? I was gonna rewrite my integration anyways but since you have one that actually works, it probably would be better to just combine both projects into a single project based on your work, since mine has issues anyways as it is one of the first home assistant integrations ive ever written haha

**Comment by sanjay900** [comment]
If your curious, here is a homebridge integration that just recently implemented the logic for detecting subscription enabled firmwares: https://github.com/EHylands/homebridge-boschcontrolpanel_bgseries

**Comment by sanjay900** [comment]
My dev panel just arrived from bosch, ill have to try hacking solution auth into your integration and see if it works
I now have a solution panel with async support!

**Comment by mag1024** [comment]
Nice, lemmy know how it goes!

I don't think adding polling support will be too hard... we can just spin off a coroutine that requests the current state of all entities / sleeps in a loop. IMO, the biggest challenge will be figuring out a code structure where the differences between the panels don't lead to ifs all over the place, but can be abstracted into some kind of model-specific polymorphic object. We'll have to play around with the code once we figure out how pervasive the divergences are.

**Comment by sanjay900** [comment]
From what i can tell, the main thing is polled vs async, and the authentication differences.

For the polled vs async, you can do a what are you with code format 0x3, and read back what commands are  supported, so we don't need to code that on a panel by panel basis thankfully, and it would be reasonable to just have reading that command jump with an if statement to an async or polled path

For the auth, i think for the moment its just solutions panels that have the weird auth requirements. I might flick bosh an email about that because it would be interesting to know why that is, and also why outputs dont work on the solutions panels (though i can finally test that now since i have a panel that i can configure without worrying about killing my home alarm system lol)

I will try your plugin as it is on the new panel though once i have it put together, it would be interesting to see if anything changed on that front, as it is quite an update to the firmware from what i can see.

**Comment by sanjay900** [comment]
@mag1024 I really did mean to get to this earlier but i got stuck with some other projects. I have just put together https://github.com/sanjay900/bosch-alarm-homeassistant and https://github.com/sanjay900/bosch-alarm-mode2
which both contain all the necessary changes to get the solution 3000 working with your addon

I have a suspicion that that will also get the AMAX going as well

**Comment by sanjay900** [comment]
Ah, arming and disarming isnt working, ill fix that

**Comment by sanjay900** [comment]
Arming and disarming is now working. I will try to do some tests with my panel that supports subscriptions later, but it would be good to know if my tests for detecting if subscriptions are available are working on your panel

**Comment by sanjay900** [comment]
So for a list of differences as i found them:
* The Arm command takes a different arm_type (ive just hardcoded this in for the moment, based on a list i found in the documentation)
* The serial number reading command doesn't exist on some of the older firmwares (this is what broke things for the AMAX i  suspect)
* Some panels don't support subscriptions
* Solution 3000 requires the different auth method
* Some panels don't support loading multiple text strings at once
* Panels that don't support subscriptions also aren't happy handling multiple commands at once, so we need to retrieve area names and point names seperately

For the serial number and subscription and the text format, we can actually ask the panel what it supports and handle support that way, instead of doing things on a panel by panel basis

For the polling, i just spun up a task for polling

For the text format stuff, I just use a different command to request a max count of items, and then i read each item up to that count.

**Comment by sanjay900** [comment]
@mag1024 
Im currently adding in support for the `triggered` state at the moment, but quick question
Is there an easy way to make home assistant update the `bosch-alarm-mode2` dependancy?

**Comment by mag1024** [comment]
Hey, very excited about all the progress!
I haven't found a way to force HA to update library dependencies -- if you figure it out, I'd love to know.
I resort to a gross hack where I have a copy of the library checked out somewhere under config/, sim-link it into the integration directory, and modify a couple of lines to force it to use the local dependency, like [so](https://github.com/sanjay900/solution3000-home-assistant/files/11527257/local-deps.txt).

**Comment by sanjay900** [comment]
I worked out I could just jump on my home assist vm, docker exec into core,
and then delete the package, but I was hoping for something I could send to
someone for testing haha, I'm tempted to just start tagging stuff

**Comment by sanjay900** [comment]
I did some digging on how home assistant handles the version numbers of packages, and it does look like its straight up impossible to get home assistant to update when using git like this, as there is no way to embed any kind of version number into the requirement in a way that you can then differenciate different builds. Pushing the package to pypi obviously works, which is what i am doing for the moment just to get it so that people can test builds.

**Comment by sanjay900** [comment]
Ill close this now as all this has been merged into your integration now!
